ykcompile/src/lib.rs

```diff
@@ -250,41 +258,99 @@ impl TraceCompiler {
+        // value is assumed to remain in rax. If we were to restore rax, we'd break that. Note that
+        // the register allocator never gives out rax for this precise reason.
+        //
+        // rdi, rsi, rdx, rcx, r8, r9, r10, r11.
+        let save_regs = [7, 6, 2, 1, 8, 9, 10, 11]
+            .iter()
+            .filter(|r| Some(**r) != dest_reg)
+            .map(|r| *r)
+            .collect::<Vec<u8>>();
+        for reg in &save_regs {
             dynasm!(self.asm
                 ; push Rq(reg)
             );
         }
 
+        let stack_index = |reg: u8| -> i32 {
+            i32::try_from(save_regs.iter().rev().position(|&r| r == reg).unwrap()).unwrap()
+        };
+
+        // Sys-V ABI dictates the first 6 arguments are passed in: rdi, rsi, rdx, rcx, r8, r9.
+        let mut arg_regs = vec![9, 8, 1, 2, 6, 7]; // reversed so they pop() in the right order.
+        for arg in args {
+            let arg_reg = if let Some(r) = arg_regs.pop() {
+                r
+            } else {
+                return Err(CompileError::Unimplemented(
+                    "out of argument registers".to_owned(),
+                ));
+            };
+
+            match arg {
+                Operand::Place(Place { local, projection }) => {
+                    if !projection.is_empty() {
+                        return Err(CompileError::Unimplemented("projected argument".to_owned()));
+                    }
+                    // Load argument back from the stack.
+                    let idx = stack_index(self.local_to_reg(*local)?);
+                    dynasm!(self.asm
+                        ; mov Rq(arg_reg), [rsp + idx * 8]
+                    );
+                }
+                Operand::Constant(c) => {
+                    dynasm!(self.asm
+                        ; mov Rq(arg_reg), QWORD c.i64_cast()
+                    );
+                }
+            };
+        }
+
         let sym_addr = TraceCompiler::find_symbol(sym)? as i64;
         dynasm!(self.asm
             // In Sys-V ABI, `al` is a hidden argument used to specify the number of vector args
-            // for a vararg call. We don't support this right now, so set it to zero. FIXME.
+            // for a vararg call. We don't support this right now, so set it to zero.
             ; xor rax, rax
             ; mov r11, QWORD sym_addr
             ; call r11
         );
 
-        // Restore registers.
-        for reg in (1..X86_86_NUM_REGS).rev() {
+        // Put return value in place.
+        if let Some(dest_reg) = dest_reg {
             dynasm!(self.asm
-                ; pop Rq(reg)
+                ; mov Rq(dest_reg), rax
             );
         }
 
-        // Put return value in place.
-        if let Some(dest) = dest {
-            let dest_reg = self.local_to_reg(dest.local)?;
+        // For tests we need this same hack as `c_enter()` uses for now.
```

Comment:
Maybe better like "To avoid breaking tests we need to the same hack as in `c_enter.`?